csrc/host_ir/jit.cpp

```diff
@@ -1245,223 +1479,45 @@ void HostIrJitImpl::registerExternalFunctions() {
-
-    const auto& kernel_inputs = compiled_kernel->kernel()->inputs();
-    const auto& kernel_outputs = compiled_kernel->kernel()->outputs();
-
-    // Create ExpressionEvaluator and bind input tensors to it
-    // This is needed for inferAndValidateAllocationSizesAndStrides to
-    // evaluate symbolic split factors and dimension extents when
-    // transforming between logical and allocation domains
-    ExpressionEvaluator expr_eval;
-    for (int64_t i = 0; i < num_inputs; ++i) {
-      if (input_tensors[i] != nullptr &&
-          i < static_cast<int64_t>(kernel_inputs.size())) {
-        expr_eval.bind(kernel_inputs[i], PolymorphicValue(*input_tensors[i]));
-      }
-    }
-
-    // Lambda to encode a tensor with allocation strides if available
-    auto encodeTensor = [&](at::Tensor* tensor, Val* tv_val) {
-      std::vector<int64_t> sizes = tensor->sizes().vec();
-      std::vector<int64_t> strides;
-      auto* tv = dynamic_cast<TensorView*>(tv_val);
-
-      // Use allocation strides if the tensor has an allocation domain
-      if (tv) {
-        if (tv->hasAllocation()) {
-          auto [alloc_sizes, alloc_strides] =
-              inferAndValidateAllocationSizesAndStrides(*tensor, tv, expr_eval);
-          strides = alloc_strides;
-        } else {
-          strides = tensor->strides().vec();
-        }
-      } else {
-        strides = tensor->strides().vec();
-      }
-
-      auto bytes = tensorToBytes(
-          PolymorphicValue(*tensor),
-          sizes,
-          strides,
-          index_type,
-          getLastDimAdjustment(tv->dtype()),
-          sizes);
-
-      arg_bytes.push_back(std::move(bytes));
-      kernel_args.push_back(arg_bytes.back().data());
-    };
-
-    // Process Inputs
-    for (int64_t i = 0; i < num_inputs; ++i) {
-      NVF_CHECK(
-          input_tensors[i] != nullptr, "Input tensor at index ", i, " is null");
-      encodeTensor(input_tensors[i], kernel_inputs[i]);
-    }
-
-    // Process Outputs
-    for (int64_t i = 0; i < num_outputs; ++i) {
-      NVF_CHECK(
-          output_tensors[i] != nullptr,
-          "Output tensor at index ",
-          i,
-          " is null");
-      encodeTensor(output_tensors[i], kernel_outputs[i]);
-    }
-
-    if (isDebugDumpEnabled(DebugDumpOption::KernelArgs)) {
-      // Convert raw tensor arrays to KernelArgumentHolder for
-      // dumpKernelArgs
-      KernelArgumentHolder input_holder;
-      for (int64_t i = 0; i < num_inputs; ++i) {
-        if (input_tensors[i] != nullptr) {
-          input_holder.push(*input_tensors[i]);
-        }
-      }
-
-      KernelArgumentHolder output_holder;
-      for (int64_t i = 0; i < num_outputs; ++i) {
-        if (output_tensors[i] != nullptr) {
-          output_holder.push(*output_tensors[i]);
+  // launch_kernel_direct function: simpler wrapper that just calls
+  // cuLaunchKernelEx All argument packing is done in LLVM IR before calling
+  // this
+  void* launch_kernel_direct_func_ptr =
```

Comment:
Given the file size is growing to >1600 lines, would you consider extracting these external functions to a different file? 

Doosn't have to be in this PR of course.